app/RouteHandlers/AddDonationHandler.php

```diff
@@ -182,4 +184,15 @@ private function newTrackingInfoFromRequest( Request $request ): DonationTrackin
 		return $tracking;
 	}
 
+	private function sendTrackingDataIfNeeded( Request $request, AddDonationResponse $responseModel ) {
+		if ( $request->get( 'mbt', '' ) !== '1' || $responseModel->getDonation()->getPaymentType() !== PaymentType::PAYPAL ) {
+			return;
+		}
+		$trackingCode = explode( '/', $request->attributes->get( 'trackingCode' ) );
+		$campaign = $trackingCode[0];
+		$keyword = $trackingCode[1] ?? '';
+		$this->ffFactory->getServerSideTracker()->setIp( $request->getClientIp() );
+		$this->ffFactory->newPageViewTracker()->trackPaypalRedirection( $campaign, $keyword );
```

Comment:
This is really ick. How about giving the IP to the tracking method? As far as I can tell there is no reason to have this mutability, much less stuffing the state in the global factory.